feat: npm based 'new bot' flow behind feature flag

pavolum opened the pull request:
## Description

This PR introduces the new creation experience work behind a hidden feature flag. It was implemented in such a way that the existing production creation experience should not be impacted whatsoever. 

This PR merges the new creation experience so far, there are various edge cases it does not yet account for, hence why it is behind a hidden feature flag. Below is a high level list of what was done

- Front and back end creation flow cloned and the copy was altered to support the new creation requirements
    - The cloned creation flow follows a 'v2' pattern where server and front end routing calls are prefaces with some 'v2' in the path
- Implementation of getReadMe server endpoint which calls npm with a given package name and returns the readMe html for that package
- Implementation of getProjTemplatesV2 server endpoint which is used to grab the npm template generators that are available in a given feed(s)
- Implementation of hidden 'New Creation Experience' feature flag which routes to the new 'new bot' flow if checked
- Added scaffolding of new runtime template for the new C# adaptive runtime, only slightly edited to ensure friendly local publish flow and build works.
- CreateProjectAsync refactored to handle npm driven template installation, building and package resolution
    - As a result the 'Conversational Core' feature will no longer work, this feature flag has been reverted to be hidden
        - Done this way to not have to support three different creation flows at once

## Task Item

fixes #5025
fixes #5026 
fixes #5488 
fixes #5502  
fixes #5498 
fixes #5489 
fixes #5566 

## Known limitations
- PVA scenarios in the new creation flow accounted for as part of this PR #5498 
- The templates instantiated in the new creation experience use the new adaptive runtime. New runtime integration is early and with this PR edge cases have not been validated, particularly around publishing. #5488  
- Detection of old runtime and bot file structure and routing to older runtime template not implemented. #5500
- Remote package acquisition does not follow agreed upon naming conventions which are TBD (component repo issue #483)
- ConversationalCore feature flag will no longer work

## Screenshots

<!---
Please include screenshots or gifs if your PR include UX changes.
--->

pavolum commented:
@benbrown pushing out draft PR for tracking. Didnt get as much time to tackle this today as id have liked but have set up the new flow behind a feature flag with cloned versions of the existing flow. Will start configuring the npm driven template feed tomorrow.

joshgummersall commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +100,79 @@ export class AssetManager {
+      log('Failed copying template to %s', dstDir);
+      throw new Error('already have this folder, please give another name');
+    }
+    await mkdirs(dstDir, (err) => {
```
Generally, async functions work in either Promise mode or callback mode, not both. [The docs](https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/ensureDir.md) indicate that if you pass a callback, a promise is not returned. Testing with the CLI seems to confirm this:

```bash
> let fsx = require('fs-extra')
undefined
> fsx.mkdirs
[Function: mkdirs]
> fsx.mkdirs('foo')
Promise { <pending> }
> fsx.mkdirs('foo', () => console.log('done'))
undefined
```

pavolum replied:
Ah good catch I changed it to use mkdirSync and added a try catch around it as seems to be the standard in other parts of the code base.

joshgummersall replied:
`await mkdirs()`, without the callback, wrapped in a `try`-`catch` would also work!

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,252 @@
+  }, [templates]);
+
+  useEffect(() => {
+    if (selectedFeed?.props?.headerText?.toLowerCase() === 'typescript') {
```
There must be a better way to access the value of the selected feed.

Also, we should pull these feeds out into a dictionary rather than define them inline here.  

`props.fetchTemplates(feeds[selectedFeed])`

benbrown replied:
maybe at least use the `itemKey` field instead of the actual label text.

Eventually these feeds will be user modifiable, so they will have a key, title, and url value.

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,252 @@
+        onDismiss={onDismiss}
+      >
+        <Pivot onLinkClick={setSelectedFeed}>
+          <PivotItem headerText="C#"></PivotItem>
```
add `itemKey` to these pivotitems

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,252 @@
+      routeToTemplate = currentTemplate;
+    }
+
+    if (option === optionKeys.createFromQnA) {
```
We need to figure out how this feature is surfaced.  This is what allows a user to start in the QNA maker mode.

benbrown replied:
This should appear in the template list even though it is not in the feed.

pavolum replied:
Ive created task #5506 to track this. I say we leave it for now and revisit it once we have an empty bot template which presumably will be used under the hood when this template is selected (instantiate empy bot template -> Route to import from qna modal and apply the uploaded KB to empty bot)

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreationFlow.tsx:
```diff
@@ -0,0 +1,192 @@
+  const currentStorageIndex = useRef(0);
+  const storage = storages[currentStorageIndex.current];
+  const currentStorageId = storage ? storage.id : 'default';
+  // const [formData, setFormData] = useState({ name: '', description: '', location: '' });
```
remove?

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/DefineConversation.tsx:
```diff
@@ -0,0 +1,329 @@
+                value={formData.description}
+                onChange={(_e, val) => updateField('description', val)}
+              />
+            </StackItem>
```
There are some fields missing from this --

we need to also collect the webapp/functions runtime choice here, as well as the primary language.  I believe there is a design for the language field in the figma files

pavolum replied:
Okay I have added those fields and nothing is done with the selection as of this PR. We can take the selections and apply them to the runtime once the runtime is ready for that, I have created a task to track that.

benbrown commented on Composer/packages/client/src/pages/home/Home.tsx:
```diff
@@ -163,9 +173,7 @@ const Home: React.FC<RouteComponentProps> = () => {
-                  TelemetryClient.track('CreateNewBotProject', { method: 'newCallToAction' });
-                  setCreationFlowStatus(CreationFlowStatus.NEW);
-                  navigate('projects/create');
+                  onClickNewBot();
```
this change will cause the telemetry to incorrectly register clicks on the call to action as clicks on the toolbar.
Maybe pass in a parameter? `onClickNewBot(method)`

pavolum replied:
Good catch, I changed it to call  the telem method from the respective onClicks for the UI components, now OnClickNewBot only calls shared functionality.

benbrown commented on Composer/packages/client/src/recoilModel/atoms/appState.ts:
```diff
@@ -271,3 +271,8 @@ export const displaySkillManifestState = atom<undefined | string>({
   default: undefined,
 });
+
+export const templateReadMeHtml = atom<string>({
```
why is this in the app state? seems like this should stay in the component.

benbrown commented on Composer/packages/client/src/recoilModel/dispatchers/storage.ts:
```diff
@@ -155,6 +156,43 @@ export const storageDispatcher = () => {
     }
   });
 
+  const fetchTemplatesV2 = useRecoilCallback(({ set }: CallbackInterface) => async (feedUrls?: string[]) => {
```
prefer all of this to be in the component itself - it doesn't need to be in app-wide state, right? It is only used in the one part of the create new wizard flow.

pavolum replied:
If we are getting rid of the template view in the home.tsx view then yeah it only needs to be in the createOptions component.

benbrown replied:
OK good point. We can leave that in place for now, but when/if the homepage gets revamped, we should move this.

benbrown commented on Composer/packages/server/src/controllers/project.ts:
```diff
@@ -518,51 +518,50 @@ function createProjectV2(req: Request, res: Response) {
-  } = req.body;
+  const { templateId } = req.body;
+
+  // todo: add back templateDir, eTag, alias from req extraction for PVA scenarios
```
lets do this before merging

pavolum replied:
I was originally planning on leaving PVA out in this PR (I am tracking its reintegration in #5498 ) but can look into it. It should be simple to put back in this v2 flow and test since it has its own isolated flow.

benbrown replied:
Would this then break PVA in the nightly builds until fixed?

pavolum replied:
It should not, there are no PVA unit tests running through this code. This PR should introduce a separate code path for creation that no existing production code paths or build blocking UTs run through unless you have the hidden feature flag checked.

pavolum replied:
@benbrown Ive restored it and done initial testing with favorable results except there are still some quirks. Getting an env to test this e2e with PVA is blocking me to iron out these quirks efficiently. I have created a task to validate the PVA flows and Tony has offered to help which should expedite the PVA integration (I believe its just one more small change needed I just cant test the damn thing effectively).

Since this will not affect anything in prod, PVA included, is it okay to iron out the kinks in its own PR so we can unblock this PR? I should I work with Tony now to get this in during this PR?

benbrown commented on Composer/packages/server/src/controllers/project.ts:
```diff
@@ -518,51 +518,50 @@ function createProjectV2(req: Request, res: Response) {
-      if (schemaUrl && !createFromRemoteTemplate) {
-        await currentProject.saveSchemaToProject(schemaUrl, locationRef.path);
-      }
+
```
is this part of the PVA related stuff?

pavolum replied:
Resolving as PVA discussion tracked elsewhere

benbrown commented on Composer/packages/server/src/controllers/project.ts:
```diff
@@ -518,51 +518,50 @@ function createProjectV2(req: Request, res: Response) {
     const id = await BotProjectService.openProject(newProjRef, user);
     // in the case of a remote template, we need to update the eTag and alias used by the import mechanism
-    createFromRemoteTemplate && BotProjectService.setProjectLocationData(id, { alias, eTag });
+    // BotProjectService.setProjectLocationData(id, { alias, eTag });
```
restore pva related stuff?

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -1,15 +1,18 @@
-import { exec } from 'child_process';
-import { promisify } from 'util';
+import path from 'path';
+// import { exec } from 'child_process';
```
remove

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -1,15 +1,18 @@
-import { promisify } from 'util';
+import path from 'path';
+// import { exec } from 'child_process';
+// import { promisify } from 'util';
```
remove

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -1,15 +1,18 @@
-import rimraf from 'rimraf';
+import { UserIdentity, FileExtensions } from '@bfc/extension';
+import { mkdirSync, readFile } from 'fs-extra';
+// import rimraf from 'rimraf';
```
remove

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -21,8 +24,9 @@ import StorageService from '../../services/storage';
 
-const execAsync = promisify(exec);
-const removeDirAndFiles = promisify(rimraf);
+// TODO: pass in working directory param to createEnv for desired location of local Yeomen Repo
```
do we still need to do this?

pavolum replied:
Resolved

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +96,78 @@ export class AssetManager {
+      return true;
     } else {
-      throw new Error('selected template has no local or external address');
+      await yeomanEnv.installLocalGenerators({ [npmPackageName]: '*' });
```
I don't see how this is ensuring that the locally installed generator is going into our composer data folder.

Perhaps setting the yeoman.cwd will do it?

pavolum replied:
I have now initialized yeomen env with the cwd set to the appData directory, it seems to be working in both web and electron context. I want to verify and see the dir itself in an electron setting. Where would that live when you run yarn start:electron:dev?

benbrown replied:
in your `~/BotFrameworkComposer` folder

benbrown commented on Composer/packages/server/src/models/bot/botProject.ts:
```diff
@@ -205,12 +205,22 @@ export class BotProject implements IBotProject {
       }
     }
 
+    if (
```
is this not something that can be done inside the yeoman template?

pavolum replied:
Good point, I just validated it and it works nicely

pavolum replied:
Kinda makes you think how much of the flow should/could be handle by the template itself. Things like the build command could be templated as well but that may be overkill
![image](https://user-images.githubusercontent.com/22845013/104507153-33115d80-559b-11eb-9512-2ee730f180cf.png)
In a larger sense the template itself could probably run things like dialog merge once it completes file generation but may be best to keep that logic in Composer if we are not needing to optimize on non Composer use of this

pavolum replied:
Yeah I guess im just realizing the power of Yeomen, cool stuff!

benbrown commented on Composer/packages/server/src/services/project.ts:
```diff
@@ -182,8 +182,8 @@ export class BotProjectService {
     }
-
-    if (!(await StorageService.checkIsBotFolder(locationRef.storageId, locationRef.path, user))) {
+    const isBotFolder = await StorageService.checkIsBotFolder(locationRef.storageId, locationRef.path, user);
```
why this change?

pavolum commented:
> make itemkey `node` to match dictionary

Okay I have resolved this and made it a bit more strongly typed as well.

benbrown commented on Composer/packages/client/src/components/ImportModal/ImportModal.tsx:
```diff
@@ -70,7 +70,8 @@ export const ImportModal: React.FC<RouteComponentProps> = (props) => {
       urlSuffix,
     };
-    let creationUrl = `/projects/create/${encodeURIComponent(source)}`;
+    // TODO: remove after testing
```
remove?

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +94,78 @@ export class AssetManager {
+      return true;
     } else {
-      throw new Error('selected template has no local or external address');
+      await this.yeomanEnv.installLocalGenerators({ [npmPackageName]: '*' });
```
we should do some error catching around this. not a blocker for now, but yeoman could fail.

benbrown replied:
since this is inside a try catch its probably ok but .. maybe some error correction

benbrown commented on extensions/runtimes/src/index.ts:
```diff
@@ -308,4 +308,174 @@ export default async (composer: any): Promise<void> => {
+        throw err;
+        return;
+      }
+      // Then, copy the declarative assets into the build artifacts folder.
```
this is probably not necessary anymore with the new runtime structure

pavolum replied:
It seems you have already commented this out, I will leave as is with commented out code

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+  ];
+
+  const onRenderRow = (props) => {
+    if (props) {
```
nit: invert this guard

```typescript
if (!props) return null;
return (<div />)
```

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+import { TemplateDetailView } from './TemplateDetailView';
+
+// -------------------- Styles -------------------- //
+
```
`detailsListContainer`, `templateDetailContainer`, `pickerContainer` don't appear to be responsive. Is this by design?

pavolum replied:
Its parent up the div chain allows for decent responsiveness at current state with these existing rules. That being said there is a task item for design iterations on this modal (responsiveness details, boilerplate readme structure, general pixel corrections). I am going to leave this as is and revisit when we do the smaller PR design iterations.

benbrown replied:
We can use the <Stack> element, it does a nice job in these cases.

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+  fetchTemplates: (feedUrls?: string[]) => Promise<void>;
+} & RouteComponentProps<{}>;
+
+export function CreateOptionsV2(props: CreateOptionsProps) {
```
I don't like this V2 convention in the class name. Can you think of alternative ways to represent a different component path than hardcoding v2 in both the directories and the classes?

pavolum replied:
Our plan is to remove the v2 convention when we make the hard shift to this creation experience prior to the build release.  That way this creation code path has the naming and file structure as the current path. 

Would there be an issue with that? I could change the 'v2' verbiage but pre build we will likely change it again to i.e. 'CreateOptions.ts' (once that file is no longer needed to support the current creation flow).

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+  const [emptyBotKey, setEmptyBotKey] = useState('');
+  const [selectedFeed, setSelectedFeed] = useState<{ props: IPivotItemProps }>({ props: { itemKey: csharpFeedKey } });
+
+  const selection = useMemo(() => {
```
`selection` of what? It's opaque. 

How about `selectedTemplate` or something more descriptive.

pavolum replied:
Handled

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+    onNext(routeToTemplate);
+  };
+
+  const tableColums = [
```
s/tableColums/tableColumns

pavolum replied:
Handled

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,268 @@
+
+  const fetchReadMe = async (moduleName: string) => {
+    try {
+      const response = await httpClient.post(`assets/templateReadme`, {
```
We're breaking the layering here by fetching data directly in the view.

Also, why is this HTTP method a post and not a get?

benbrown replied:
I thought it would be appropriate to have this data being fetched at the component level since it is not used anywhere else. Would you have it live in the app state?

pavolum replied:
I agree that it should be in app state, ill add it back there

benbrown replied:
¯\_(ツ)_/¯

pavolum replied:
Its like the whole division of business logic and presentational logic.

Ive also been on teams that ban local state in favor of central. Like even if a component is the only one that needs that data it may not be the case down the line so always place state in a central repo like recoil or redux. 

Thats a little overkill imo but I do like the division of business logic and presentational logic aspect.

pavolum replied:
Actually I am going to leave this as is. We need to settle the ideal read me/template detail acquisition story. If we are no longer going down the readme grabbing route then it does not make sense to move this to app state since that will potentially be throw away code.

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/DefineConversation.tsx:
```diff
@@ -0,0 +1,369 @@
+
+const MAXTRYTIMES = 10000;
+
+interface DefineConversationFormData {
```
this shape should be better typed, and our convention is to use types over interfaces

pavolum replied:
Handled

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/DefineConversation.tsx:
```diff
@@ -0,0 +1,369 @@
+  preserveRoot?: boolean; // identifier that is used to determine ay project file renames upon creation
+}
+
+interface DefineConversationProps
```
our convention is to use types over interfaces

pavolum replied:
Handled

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/DefineConversation.tsx:
```diff
@@ -0,0 +1,369 @@
+    }
+  };
+
+  const getDefaultName = () => {
```
was this function copied from somewhere? it seems out of place

pavolum replied:
Yes this is from the original DefineConversation.tsx and slightly altered to provide a valid default project name to prepopulate the name field in the UI. There is a good bit of redundant code between the two flows which is intended since they will only live side by side for a short period of time.

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/DefineConversation.tsx:
```diff
@@ -0,0 +1,369 @@
+      }
+
+      onSubmit(
+        {
```
does prettier format this object parameter like it's checked-in? why not `{ ...dataToSubmit },`

pavolum replied:
Handled

cwhitten commented on Composer/packages/client/src/components/CreationFlow/v2/TemplateDetailView.tsx:
```diff
@@ -0,0 +1,16 @@
+};
+
+export const TemplateDetailView: React.FC<TemplateDetailViewProps> = (props) => {
+  return <div className="content" dangerouslySetInnerHTML={{ __html: props.readMe }}></div>;
```
We really should not `dangerouslySetInnerHTML`

pavolum replied:
Yeah Im not a big fan of it for the xss attack vulnerability although we control the src it pulls the html from. 

We would have to rethink how we grab the template detail view. I like the idea of grabbing the readme as it makes template development easier and it sticks with the npm standard. 

Alternatively, the feed sends metadata that could pass the core info of the template (i.e. supported languages, description, etc) to Composer for rendering. It would make visual styling easier and more consistent then pulling readMe markdown.

@benbrown what do you think the meta data config would be in package.json of the template. So instead of populating a readme they would populate a handful of fields in the package.json of their template package.

pavolum replied:
Resolving as the new readMe approach has been added that takes call logic out of the component layer, adds readme to app state and does not dangerouslyinjecthtml

cwhitten commented on Composer/packages/server/src/controllers/asset.ts:
```diff
@@ -2,8 +2,11 @@
 // Licensed under the MIT License.
```
A general ask for this file -- is all the data fetching in the controller layer the right thing?

pavolum replied:
Okay I have exported most of these additions to AssetManager. I have left the readme function here since your other comment may lead to the removal of this technique.

cwhitten commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +93,86 @@ export class AssetManager {
+        await this.instantiateRemoteTemplate(generatorName, dstDir, projectName);
       }
+
+      ref.path = ref.path + `/${projectName}`;
```
nit: interpolation over concatenation

pavolum replied:
Handled

cwhitten commented on extensions/azurePublish/src/node/luisAndQnA.ts:
```diff
@@ -12,8 +12,8 @@ import { AzurePublishErrors } from './utils/errorHandler';
 const readdir: any = promisify(fs.readdir);
-
-const botPath = (projPath: string) => path.join(projPath, 'ComposerDialogs')
+// const botPath = (projPath: string) => path.join(projPath, 'ComposerDialogs')
```
kill this if unneeded?

pavolum replied:
I didnt make this change, @benbrown is it okay to remove this? Note changes to this file will affect the existing creation flow

pavolum replied:
I have removed it but will leave this comment to confirm with @benbrown

cwhitten reviewed (dismissed):
general comments: there is random formatting changes across a host of files adding things like whitespace. why?

there are no new tests are part of this. it seems like there should be.

it would be beneficial to improve the typing across the board

why are we checking in package-lock.json's?

pavolum commented:
> general comments: there is random formatting changes across a host of files adding things like whitespace. why?
**Must be an auto formatter mismatch that altered those files, I have reverted these changes.** 

> there are no new tests are part of this. it seems like there should be.
**Small aspects of the flow may be subject to change and this is behind a hidden feature flag which is why we are waiting to develop UTs. That being said if its a hard block for merges to main even if the code is hidden from end users, then I can tackle an initial set now and alter them as/if the code changes. @benbrown what do you think, should UTs be prioritized before pushing this PR?**

> it would be beneficial to improve the typing across the board
**I have improved some of the typing you explicitly called out. I largely stuck with the patterns implemented in the existing creation experience and the standards their seem to be outdated (interfaces are used over types, etc) so I have updated those as well. Are there any other specific typing issues that you would want to see resolved before this is pushed?**

> why are we checking in package-lock.json's?
**I have removed this**

cwhitten commented:
> I can tackle an initial set now and alter them as/if the code changes.

Sounds good! 

> I largely stuck with the patterns implemented in the existing creation experience and the standards their seem to be outdated

I understand. The downside of copying a bunch of code and checking it in is you now own the copy. I'll take a closer look at the types soon.

pavolum commented:
Some things are being brought up here that are good example of the need to clear up the following devops question that is introduced with the hidden feature flag concept.

**Should we be working on a separate feature branch until this is at its first e2e acceptable state and then PR or should we push incrementally to main and hide the incomplete (working but no UT, end designs, etc) code behind a hidden feature flag?**

This PR assumes the latter which is a new approach then what we are used to. 

Two options for feature work:
- Large features are worked on in their own feature branch and not pushed to main until they are in shipping condition (UTs developed, final designs implemented, etc.)
  - Cons:
    - Need to switch to and build feature branches to try new features
    - Potential for larger rifts from main if you do not update feature branches regularly
  - Pros:
    - Everything in main is in its final state and code is guaranteed to not go unreviewed
- Large feature work incrementally pushed to main behind a hidden feature flag
  - Pros:
    - Regular integration with main leads to less branch rifting 
    - Other devs can easily turn on hidden feature flag work in main to test in development work.
  - Cons:
    - Potential for unimplemented code (i.e first incremental push to main has not UTs or final designs and its on the word of the dev to actually implement those at a later PR prior to shipping)

pavolum commented:
Oops accidentally deleted a quote reply comment that was outdated and it seems to have deleted the original, will try to recover.

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreateOptions.tsx:
```diff
@@ -0,0 +1,259 @@
+          }}
+        >
+          <PivotItem headerText="C#" itemKey={csharpFeedKey}></PivotItem>
+          <PivotItem headerText="Typescript" itemKey={nodeFeedKey}></PivotItem>
```
Before we ship, we can kill this pivot since we will not have Typescript templates. Til then let's leave it to test underyling features.

pavolum replied:
Okay I will disable it as an option so users wont be able to click it

pavolum replied:
Hmm it seems that UI element is not disableable. I have removed it for now so there is only the C# tab. Will iterate with design for a better short term solution if needed. (iterating with design task is created and will be handled in its own PR)

benbrown commented on Composer/packages/client/src/components/CreationFlow/v2/CreationFlow.tsx:
```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// TODO: Remove path module
```
why this todo?

pavolum replied:
Removed

benbrown commented on Composer/packages/client/src/constants.ts:
```diff
@@ -379,3 +390,25 @@ export const triggerNotSupportedWarning = () =>
+export const feedDictionary: { [key in FeedName]: string } = {
+  firstPartyCsharp:
+    'https://registry.npmjs.org/-/v1/search?text=conversationalcore&size=100&from=0&quality=0.65&popularity=0.98&maintenance=0.5',
+  firstPartyNode:
```
we'll want to remove/disable the JS feed before shipping

pavolum replied:
Handled

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +97,86 @@ export class AssetManager {
+
+      if (remoteTemplateAvailable) {
+        await this.instantiateRemoteTemplate(generatorName, dstDir, projectName);
       }
```
do we need an else here?

benbrown replied:
maybe we should return true / false from installRemoteTemplate instead of using throw/catch?

lets pick one instead of both.

pavolum replied:
Handled

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +97,86 @@ export class AssetManager {
+  ): Promise<boolean> {
+    this.yeomanEnv.cwd = dstDir;
+
+    log('About to instantiate a template!', dstDir, generatorName, projectName);
```
does awaiting this work?


normally when there is a callback, it doesn't work as a promise.

We should double check on this - otherwise, wrap this in a Promise and resolve inside the callback.

pavolum replied:
Just ran a test, awaiting it does work :D. It runs the callback prior to continuing the function and returning 'true'

benbrown commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -234,4 +267,56 @@ export class AssetManager {
   }
+
+  private getFeedType(): FeedType {
+    // TODO: parse through data to detect for npm or nuget package schema and return respecive result
```
I have code that does this in package manager.

pavolum replied:
Ah thats right, I remember you mentioning that a while back. How shall we proceed? Want to send me the deets to that implementation and I can make the change to use that function?

cwhitten commented on Composer/packages/server/package-lock.json:
```diff
@@ -0,0 +1,1780 @@
+{
```
we should also remove this

pavolum replied:
Man, I accidentally run one 'npm' command instead of 'yarn' and these files sneak in there. Ill remove now.

cwhitten commented on Composer/packages/server/src/controllers/__tests__/project.test.ts:
```diff
@@ -172,6 +172,49 @@ describe('should get recent projects', () => {
+  });
+
+  it('should create a new project', async () => {
+    // await ProjectController.createProjectAsync(mockReq, '123');
```
is this intended to be commented out?

pavolum replied:
Ah thats remnants of an old test, removing now.

cwhitten commented on Composer/packages/server/src/controllers/asset.ts:
```diff
@@ -16,6 +18,57 @@ async function getProjTemplates(req: Request, res: Response) {
+
+    // Grab templates from FeedURls
+    if (feedUrls) {
+      // TODO: Current assumption is that the feed is an npm feed with given result structure (https://registry.npmjs.org/-/v1/search?text=docker&size=100&from=0&quality=0.65&popularity=0.98&maintenance=0.5)
```
what's the future plan here with this TODO?

pavolum replied:
Ive removed the TODO, the work needed for that revolved around documentation tasks that have been created and are being tracked. Essentially we just need to let folks know through the docs the standard expected structure of the template feeds once we allow custom feeds for templates (post R12). 
The expected structure is set by the respective package manager (npm) so we are not creating our own feed schema or anything, we are sticking to the format that  npm dictates to not uneccisarily reinvent the wheel. 

I have removed the TODO.

cwhitten commented on Composer/packages/server/src/controllers/asset.ts:
```diff
@@ -16,6 +18,57 @@ async function getProjTemplates(req: Request, res: Response) {
+        message: 'missing module name on request',
+      });
+    } else {
+      const moduleURL = 'https://registry.npmjs.org/' + moduleName;
```
nit: interpolation over concatenation

pavolum replied:
Handled

cwhitten commented on Composer/packages/server/src/models/asset/assetManager.ts:
```diff
@@ -92,58 +97,88 @@ export class AssetManager {
+
+    log('About to instantiate a template!', dstDir, generatorName, projectName);
+    await this.yeomanEnv.run([generatorName, projectName], {}, () => {
+      console.log('DONE');
```
Is this necessary?

pavolum replied:
Console.log is a bit invasive. I have changed to to use the log() function.

cwhitten commented on extensions/packageManager/src/client/Library.tsx:
```diff
@@ -294,16 +295,20 @@ const Library: React.FC = () => {
+    if (!isLoadingInstalled) {
+      setIsLoadingInstalled(true);
+      try {
+        updateInstalledComponents([]);
```
why does this function invocation need to accept an empty array?

benbrown replied:
That is resetting the state to an empty list while it loads...

cwhitten commented:
> ConversationalCore feature flag will no longer work

Other work streams are relying on this feature flag. Composer Adapters for one. Can you make sure to connect with Ben Y on the appropriate way to guard UX associated with the VA scenarios?

pavolum commented:
> > ConversationalCore feature flag will no longer work
> 
> Other work streams are relying on this feature flag. Composer Adapters for one. Can you make sure to connect with Ben Y on the appropriate way to guard UX associated with the VA scenarios?

Good to know, Ill ping him now!

pavolum commented:
> > > ConversationalCore feature flag will no longer work
> > 
> > 
> > Other work streams are relying on this feature flag. Composer Adapters for one. Can you make sure to connect with Ben Y on the appropriate way to guard UX associated with the VA scenarios?
> 
> Good to know, Ill ping him now!

I spoke with Ben Y and we have a plan in place. Once this gets into main he will pull latest and make it such that the adapter feature is only shown if the NEW_CREATION_FLOW feature flag is checked. That way when I do the smaller PR that gets rid of the REMOTE_TEMPLATE_CREATION feature flag then his branch wont break when he pulls latest.
